airflow/decorators/assets.py

```diff
@@ -0,0 +1,159 @@
+            asset_names.append(self._definition_name)
+
+        if asset_names:
+            self._active_assets = _fetch_active_assets_by_name(asset_names)
+        return dict(self._iter_kwargs(context))
+
+
+def _handle_self_argument(func: types.FunctionType) -> types.FunctionType:
+    @wraps(func)
+    def wrapper(_self: Any, *args: Sequence[Any], **kwargs: dict[str, Any]) -> Any:
+        return func(_self, *args, **kwargs)
+
+    return cast(types.FunctionType, wrapper)
+
+
+@attrs.define(kw_only=True)
+class AssetDefinition(Asset):
+    """
+    Asset representation from decorating a function with ``@asset``.
+
+    :meta private:
+    """
+
+    function: types.FunctionType
+    schedule: ScheduleArg
+
+    def __attrs_post_init__(self) -> None:
+        parameters = inspect.signature(self.function).parameters
+        if "self" in parameters:
+            self.function: types.FunctionType = _handle_self_argument(cast(types.FunctionType, self.function))  # type: ignore[assignment]
+
+        with DAG(dag_id=self.name, schedule=self.schedule, auto_register=True) as dag:
+            _AssetMainOperator(
+                task_id="__main__",
+                inlets=[
+                    AssetRef(name=inlet_asset_name)
+                    for inlet_asset_name in parameters
+                    if inlet_asset_name not in ("self", "context")
+                ],
+                outlets=[self.to_asset()],
+                python_callable=self.function,
+                definition_name=self.name,
+                uri=self.uri,
+            )
+
+        # TODO: Currently this just gets serialized into a string.
+        # When we create UI for assets, we should add logic to serde so the
+        # serialized DAG contains appropriate asset information.
+        dag._wrapped_definition = self
+
+        DAG.bulk_write_to_db([dag])
+
+    def to_asset(self) -> Asset:
+        return Asset(
+            name=self.name,
+            uri=self.uri,
+            group=self.group,
+            extra=self.extra,
+        )
+
+    def serialize(self):
+        return {
+            "uri": self.uri,
+            "name": self.name,
+            "group": self.group,
+            "extra": self.extra,
+        }
+
+
+@attrs.define(kw_only=True)
+class asset:
+    """Create an asset by decorating a materialization function."""
+
+    schedule: ScheduleArg
+    uri: str | ObjectStoragePath | None = None
+    group: str = attrs.field(
+        kw_only=True,
+        default="",
+        validator=[attrs.validators.max_len(1500), _validate_identifier],
+    )
+    extra: dict[str, Any] = attrs.field(factory=dict)
+
+    def __call__(self, f: Callable) -> AssetDefinition:
+        if (name := f.__name__) != f.__qualname__:
+            raise ValueError("nested function not supported")
+        if name == "self" or name == "context":
+            raise ValueError(f"prohibited name for asset: {name}")
```

Comment:
Doing `Asset(name="self")` (and `context`) directly should also be banned. We should move these checks to `Asset` instead.

Reply:
Sounds good. will do.